invoice/src/main/java/org/killbill/billing/invoice/usage/SubscriptionUsageInArrear.java

```diff
@@ -180,6 +193,21 @@ public UsageKey apply(final Usage input) {
                 }
             }
         }
+
+        // Sanity check: https://github.com/killbill/killbill/issues/1275
+        for (final RawUsage rawUsage : rawSubscriptionUsage) {
```

Comment:
This code could probably be moved inside `ContiguousIntervalUsageInArrear#getRolledUpUsage` where we already iterate across all `rawSubscriptionUsage`  elements?

Reply:
I agree there is a bug in the scenario you mentioned above (I was able to reproduce it in a test case), but I'm struggle coming up with a fix.

First, to move the logic into `getRolledUpUsage`, this logic has to be moved to: https://github.com/killbill/killbill/blob/19b338ffba1e8998ff80ac7d894a0f0672e7fe80/invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalUsageInArrear.java#L435-L443

If I do so though, as I'm iterating through the usage data in `getRolledUpUsage`, it's hard the filter the recorded usage by unit type as it's unclear if the unknown recorded data is either not defined in the catalog, or simply for another usage section (one billing event maps to several `ContiguousIntervalUsageInArrear`).

The only solution I can find would be to move my logic inside the `for` loop of `computeInArrearUsageInterval` (meaning I would need to expose `getRolledUpUsage` publicly and call it there to filter the usage by date).

Might be worth for you refreshing your memory on this and talk through it in more details, in case I'm missing something...

Reply:
I looked at it, and i think one issue is that code from `0.20` and `0.22` has diverged quite a bit: In particular, i had added a mechanism to link a `transitionTime` with its `billingEvent`: `https://github.com/killbill/killbill/commit/48681d8821db97a0952482e6abf8a308dea2f918#diff-3e43568d6c03fc887b25fd5883cb0c3aR99`

So, what i had in mind was:
1. Compute the `allSeenUnitTypes` per `billingEvent` (similar to `toBeClosed`)
2. Pass the `allSeenUnitTypes`  to `interval.addBillingEvent(event)` and leverage the `TransitionTime` structure to also attach the `allSeenUnitTypes`

=> At this point in `ContiguousIntervalUsageInArrear` we have the mapping between a given 
`transition` -> `billingEvent` -> `allSeenUnitTypes` and we should be able as we go through the `RolledUpUsage` and check wether we see entries contains unit that are not part of the `allSeenUnitTypes`  (for that specific `billingEvent`)

But... this code is not available on `0.20` so that makes things a bit more complex...

Reply:
> => At this point in `ContiguousIntervalUsageInArrear` we have the mapping between a given `transition` -> `billingEvent` -> `allSeenUnitTypes` and we should be able as we go through the `RolledUpUsage` and check wether we see entries contains unit that are not part of the `allSeenUnitTypes` (for that specific `billingEvent`)

This is not enough though because of the `filterInputRawUsage` I mentioned above: each interval knows about the billing events but not about other intervals (unknown unit type could be caused either by a missing entry in the catalog or because it is owned by another interval -- typically when you have both capacity and consumable for the same plan).

I tried a few approaches and finally settled on a mix of backport of your patch + removing the logic `filterInputRawUsage`. To be reviewed...